include/pybind11/pytypes.h

```diff
@@ -1320,9 +1339,14 @@ class sequence : public object {
 class list : public object {
 public:
     PYBIND11_OBJECT_CVT(list, object, PyList_Check, PySequence_List)
-    explicit list(size_t size = 0) : object(PyList_New((ssize_t) size), stolen_t{}) {
+    explicit list(ssize_t size = 0) : object(PyList_New(size), stolen_t{}) {
```

Comment:
Should we have asserts for a negative ssize?

Reply:
What does `PyList_New` to in that case?
Do we already have a unit test for the max value of `size_t`?
I'd answer the question from that angle.

Reply:
Good point.

Reply:
PyList_New explicitly checks for negative sizes and errors on them: https://github.com/python/cpython/blob/4ceec495598e78f0776dd46d511dcc612a434dc3/Objects/listobject.c#L142
So in that sense (I didn't actually realize this before) I'd argue that supporting size_t inputs is actually less correct than supporting ssize_t, because the former suggests that e.g. (SIZE_T_MAX-1) would be a valid input whereas it is not.

Reply:
I agree with that. I think it's best to simply replace all `size_t` with `ssize_t`, without adding new overloads, and noting the change in the upgrade guide.

Just that will leave people out in the cold when they get the `-Wnarrowing` diagnostic, but if we also add something like a `py::safe_as_ssize_t(T)` helper in this PR we could tell them to use that. I think (very old) boost code has safe conversions from any integral type to a narrower integral type. Probably there are many solutions out there, but the boost one is likely to get all corner cases right. Or maybe there is even something in the C++ standard library? I don't know and didn't look.

In summary:
* pybind11 will be safer.
* users will have a choice, hard cast (they need to take responsibility explicitly), or the safe helper with a runtime check.

Reply:
:tada:

I can add `py::safe_as_ssize_t` if you prefer (well, as you say, it may be worth finding a good implementation somewhere, which I don't have right now), but once again, note that if the input was between SSIZE_T_MAX and SIZE_T_MAX, then we were *already* going to get a C++ exception thrown at that point (because PyList_New would have returned NULL, which pybind11 already catches and converts to a C++ exception), so I don't think another helper just so that the exception gets thrown *before* the call to PyList_New rather than after adds much?